### ISSUE TITLE
First time WebUSB download takes a while

### ISSUE THREAD

**Issue by Jaqster** [microsoft/pxt-microbit#5093]
Not sure if there's anything that can be done to improve performance here, but the first time you do a WebUSB program download, it's taking 5-10 seconds to download :-(

![image](https://user-images.githubusercontent.com/22108954/234989338-7e58a0c3-5e73-4124-af10-98f0434119fe.png)

**Comment by jaustin** [comment]
5-10 seconds is really fast. I'd expect first time download on a slow computer to be around 30 seconds. Some of the machines we tried at BETT were up to 1 minute.

It appears to be CPU-related - if you try throttling your CPU in dev tools it slows down.

We thought about this in the Python editor and decided on the following Ux that has gone down quite well:
1. tell people that first time flash will take longer, and give them a progress bar
2. promise them quicker flashes after that

<img width="611" alt="image" src="https://user-images.githubusercontent.com/147113/234997087-61bd9449-2b3e-49da-8d30-13ea67f9a6ee.png">

![image](https://user-images.githubusercontent.com/147113/234997326-de61c0ee-1acf-4125-9ffd-6a0dc2171c25.png)

**Comment by Jaqster** [comment]
Yikes!  I think a 1-minute download is not acceptable.  Is there anything we can do to improve performance for the first-time download?

**Comment by jaustin** [comment]
I think something to ask @mmoskal - I guess because it's CPU bound on the host, hopefully. I'm not sure if anyone has profiled the web USB code to see where the cycles are spent?

**Comment by jaustin** [comment]
Some benchmarking here.
Using 4x CPU throttling on my M1 Macbook Pro I see ~35 seconds for a web USB flash from MakeCode (replacing MicroPython) and the same for replacing MakeCode when I've just added datalogging extension. (FWIW, this is about half the time the Python Editor takes with the same 4x slow down, so I think MakeCode's approach is pretty good here.

There are two ways to 'full flash' a micro:bit - one is to send the whole hex (either universal, or the intel hex for the specific version of the hardware) and let the micro:bit do a full erase and flash. 
The other is to use the partial flashing mechanism over the whole of flash. 

If using the former method it will be quicker to only send the intel hex, not a universal hex. By the way the micro:bit is behaving, I think MakeCode's using the latter method anyway.

Finally, there definitely seems to be a few things that happen that take some time before the flashing even starts (measured for me by when the orange LED starts to blink) - especially if you've just added an extension. This is much more visible with throttling on.

**Comment by abchatra** [comment]
@jwunderl there is progress indicator in skillmap->tutorial loading screen. Can we reuse that?

**Comment by microbit-carlos** [comment]
Have kind of timings do you get on the same machine when setting the Chrome/Edge developer tools CPU throttling to x6?

**Comment by jwunderl** [comment]
Closing as I added progress indicator https://github.com/microsoft/pxt-microbit/pull/5200 

I also poked around the rest of the webusb stack quite a bit for if there was anything we can do to trim off any time and I didn't find anything in particular; I put up a PR swapping from the explicit promise chain to async / await loops in case there was any time we could snag out of it (e.g. potentially browser optimizations could treat it better than a long tail recursive chain) but I didn't get a noticeable amount of time -- https://github.com/microsoft/pxt-microbit/pull/5206 has a before / after build to try, under 4x cpu throttling it seemed to be ~ .5-1 second faster for a full flash but it was too inconsistent to confirm / within the margin of error (e.g. range for a full flash was around 38-44 seconds in those circumstance, mostly just awaiting webusb calls). If anyone does have a device where flashing runs particularly slowly, and sees a difference between the two builds, it'd be very useful to know!
